### ISSUE TITLE
TLS Certificate Viewer

### ISSUE THREAD

**Issue by digitarald** [devtools-html/planning#1]
# Intro

This document will outline how to make Certificate information available in the developer tools for security auditing and debugging.

# Target Audience

Various functions in DevOps at large (backend developers, security engineers, quality assurance, infrastructure, site reliability, etc) – who need to audit certificates and debug issues.

# Problem Statement

The current placement in the _Page Info_ dialog is showing developer-focussed information that is more relevant in DevTools. Firefox's long-term plan is to remove the page info dialog as most information is available in the awesomebar doorhanger.

# Functional Requirements

April King already created a [certificate viewer extension](https://github.com/april/certainly-something) that would be great to have as devtools panel.

Requirement | Job Story | Priority | Notes
-- | -- | -- | --
View Certificate | When I load a secure site, I want to inspect the certificate information for the connection, so that I can review the certificate details | Must have | 
Download Certificate | When inspecting a certificate, I want to download the certificate, so that I can analyze it with other tools, share it or save it for later | Must have | Chrome does this through the OS viewer.

# Technical Requirements

Questions | Outcome
------------ | -------------
New Security Panel or App Panel? | Security panel can be worked on faster. App panel risks to become overloaded. *Needs decision.*
Show Certificates for 3rd-party resources? | Chrome shows a hierarchy of origins to navigate. *Needs decision.*

**Comment by digitarald** [comment]
@nightwarrior-xxx one simple first step would be to make the existing cert viewer extension compatible with devtools tabs, so its usable within devtools and doesn't add a new icon to the Firefox toolbar. That would make it easier to migrate as second step.

**Comment by nightwarrior-xxx** [comment]
@digitarald That could be the first step .But then how would user can see or download the certificates.
We have to give them an *download certificate button* to download certificates.
Moreover I would love to work on it.

**Comment by digitarald** [comment]
@nightwarrior-xxx the basic idea would be to show the certificate summary that the extension provides in a devtools tab.

I'll make a mock up as well for the requirements doc and add these notes.

As reference, the in-tab summary that the extension provides:

![image](https://user-images.githubusercontent.com/8599/46305510-7f6d3680-c566-11e8-89b6-9d6e7eecfd43.png)

The extension also has the option to download:

<img width="464" alt="image" src="https://user-images.githubusercontent.com/8599/46305416-34532380-c566-11e8-83d6-07e8210e3ae2.png">

**Comment by april** [comment]
Happy to help in any way that I can with this!  (I wrote the extension above)

**Comment by digitarald** [comment]
@april also shared this awesome screenshot that shows the extension working in Firefox on Android:

![image](https://user-images.githubusercontent.com/8599/46306287-ad537a80-c568-11e8-8449-5e474b32ae9f.png)

**Comment by nightwarrior-xxx** [comment]
@digitarald Can we start this project/task or we should wait for your mock up?
Regards,
@nightwarrior-xxx

**Comment by violasong** [comment]
Just to tie things together: It seems this might be related to [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=396441)?

**Comment by digitarald** [comment]
> Just to tie things together: It seems this might be related to this bug?

Yes, this 11 year old bugs seems relevant (great bug archeology!) and there are recent plans around sunsetting the Page Info dialog as it is riddled with XUL.

@nightwarrior-xxx, sorry for not responding quickly as I was out on a long weekend. The UI April did should work great in a panel as it is responsive; so we won't need much UX help for the MVP.

**Comment by violasong** [comment]
Cool :). I'll update that bug to point to this bug. (A month ago I had chatted with g.rossolini who was potentially interested in helping with the UX for this bug).

**Comment by digitarald** [comment]
Also see the related [XUL Tree Migration](https://docs.google.com/document/d/1lrWIAKTGfyO8_FaAScUc9V57jyZdM-odig00MmFh9HA/edit?ts=5bb86fd6#heading=h.yet60fxp64q8) doc here, as the existing Certificate viewer is blocking XUL Migration.

**Comment by elFarto** [comment]
> > Just to tie things together: It seems this might be related to this bug?
> 
> Yes, this 11 year old bugs seems relevant (great bug archeology!) and there are recent plans around sunsetting the Page Info dialog as it is riddled with XUL.

I'm not sure that bug has anything to do with this issue. That bug is specifically for TLS client authentication (that is allowing the user to send a TLS certificate to the server).
